### PR TITLE
python3Packages.txtai: 7.3.0 -> 7.4.0 plus fixes

### PR DESCRIPTION
**Pull request by sarahec**
1. Bump to 7.4.0
2. Re-enable building on Python 3.12. Was due to the python imports check running code that wrote to a cache dir. Called `mktemp -d` at the end of the `distPhase`. (! Feels hacky, but is apparently the way.)
3. Re-enabled `checkPhase` for all.
4. Added `ann` optional dependency.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [x] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by sarahec** [comment]
Result of `nixpkgs-review pr 347622` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>4 packages built:</summary>
  <ul>
    <li>python311Packages.txtai</li>
    <li>python311Packages.txtai.dist</li>
    <li>python312Packages.txtai</li>
    <li>python312Packages.txtai.dist</li>
  </ul>
</details>

**Comment by sarahec** [comment]
@ofborg build python312Packages.txtai

File: pkgs/development/python-modules/txtai/default.nix [hunk 1]
```diff
@@ -176,23 +190,49 @@ buildPythonPackage {
...
+    "--deselect=test/python/testembeddings.py"
+    "--deselect=test/python/testgraph.py"
+    "--deselect=test/python/testapi/testembeddings.py"
+    "--deselect=test/python/testapi/testpipelines.py"
+    "--deselect=test/python/testapi/testworkflow.py"
+    "--deselect=test/python/testdatabase/testclient.py"
+    "--deselect=test/python/testdatabase/testduckdb.py"
+    "--deselect=test/python/testdatabase/testencoder.py"
+    "--deselect=test/python/testworkflow.py"
+  ];
```
Comment (GaetanLepage) [inline pkgs/development/python-modules/txtai/default.nix:225]: Can't we use `disabledTests` for those ?

File: pkgs/development/python-modules/txtai/default.nix [hunk 2]
```diff
@@ -176,23 +190,49 @@ buildPythonPackage {
...
 
-  pythonImportsCheck = [ "txtai" ];
-
-  # some tests hang forever
-  doCheck = false;
-
-  preCheck = ''
-    export TRANSFORMERS_CACHE=$(mktemp -d)
+  # The Python imports check runs huggingface-hub which needs a writable directory
+  postInstall = ''
```
Comment (GaetanLepage) [inline pkgs/development/python-modules/txtai/default.nix:194]:
```suggestion
  postInstall = ''
```
Usually, we would do that in `preCheck`.

**Review by GaetanLepage — CHANGES_REQUESTED** [review]
Also:
- Try to replace `format = "setuptools";` with `pyproject = true;` (this might require adding `build-systems = [ setuptools ];`
- `propagatedBuildInputs` -> `dependencies`

File: pkgs/development/python-modules/txtai/default.nix (hunk 1 above)
Comment (sarahec) [reply pkgs/development/python-modules/txtai/default.nix:225]: Sadly, no. The failures happen when initializing the classes holding the tests (no way to list those in disabledTests).  Also, `disabledTestPaths` didn't work. I put them here after two hours of trying with disabledTestPaths.

File: pkgs/development/python-modules/txtai/default.nix (hunk 2 above)
Comment (sarahec) [reply pkgs/development/python-modules/txtai/default.nix:194]: It doesn't work in preCheck, unfortunately. Inspecting the build log shows that postCheck runs well before the python imports check. I spent quite a while finding the right place to set this value.

Comment (sarahec) [reply pkgs/development/python-modules/txtai/default.nix:194]: Here's what happens when I set this value in `postInstall`:

```sh
Running phase: installCheckPhase
no Makefile or custom installCheckPhase, doing nothing
Running phase: pythonCatchConflictsPhase
Running phase: pythonRemoveBinBytecodePhase
Running phase: pythonImportsCheckPhase
Executing pythonImportsCheckPhase
Check whether the following modules can be imported: txtai
Running phase: pytestCheckPhase
```

It works.

Comment (sarahec) [reply pkgs/development/python-modules/txtai/default.nix:194]: However, it works in `preInstall` which is in line with what the pore/post hooks should do.

Comment (sarahec) [reply pkgs/development/python-modules/txtai/default.nix]: Done as postInstall -> preInstall.

**Comment by sarahec** [comment]
@GaetanLepage that's your changes added, except where they couldn't be. I also added comments to preInstall and pytestCheckFlags documenting why this is set up as it is.

**Review by GaetanLepage — APPROVED** [review]
LGTM !

**Comment by GaetanLepage** [comment]
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 347622`

---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 4 packages built:</summary>
  <ul>
    <li>python311Packages.txtai</li>
    <li>python311Packages.txtai.dist</li>
    <li>python312Packages.txtai</li>
    <li>python312Packages.txtai.dist</li>
  </ul>
</details>

---
### `aarch64-linux`
<details>
  <summary>:x: 2 packages failed to build:</summary>
  <ul>
    <li>python312Packages.txtai</li>
    <li>python312Packages.txtai.dist</li>
  </ul>
</details>
<details>
  <summary>:white_check_mark: 2 packages built:</summary>
  <ul>
    <li>python311Packages.txtai</li>
    <li>python311Packages.txtai.dist</li>
  </ul>
</details>

---
### `x86_64-darwin`
<details>
  <summary>:fast_forward: 4 packages marked as broken and skipped:</summary>
  <ul>
    <li>python311Packages.txtai</li>
    <li>python311Packages.txtai.dist</li>
    <li>python312Packages.txtai</li>
    <li>python312Packages.txtai.dist</li>
  </ul>
</details>

---
### `aarch64-darwin`
<details>
  <summary>:white_check_mark: 4 packages built:</summary>
  <ul>
    <li>python311Packages.txtai</li>
    <li>python311Packages.txtai.dist</li>
    <li>python312Packages.txtai</li>
    <li>python312Packages.txtai.dist</li>
  </ul>
</details>

**Comment by GaetanLepage** [comment]
> `nixpkgs-review` result

`aarch64-linux` failure is caused by `faiss` being broken (fixed in #347735).

**Pull request merged**